### DISCUSSION TITLE
Create route that has @ in the url

### DISCUSSION THREAD

**Discussion by litlatte** [vercel/next.js#50753] [App Router] [answered]
Is there a way to create a route that has a prefix (a character or string before the dynamic part) in the url when using the new app folder ?
e.g. https://website.com/@hello
where @ is our prefix

By default now this gets interpreted as a slot

**Comment by devjiwonchoi** [comment]
You can use [Dynamic Routes](https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes) and [Middleware](https://nextjs.org/docs/app/building-your-application/routing/middleware) to resolve this.

Set your `/@hello` pages inside a Dynamic Route segment such as `/app/[slug]/page.tsx` or `/app/[@hello]/page.tsx`,

![Screenshot 2023-06-04 at 10 28 20 PM](https://github.com/vercel/next.js/assets/120007119/c1c11b88-685a-4a61-b390-e60876e0fc85)

and modify your `middleware` to check the path name starts with `/@` and `init the Response`.

Example:

```js
// middleware.ts

import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'

export function middleware(request: NextRequest) {
  if (request.nextUrl.pathname.startsWith('/@')) {
    return NextResponse.next()
  }
}
```

Note that the `middleware` file should be located in the root segment of your project as `/src/middleware.ts` or `/middleware.ts`.

> Use the file middleware.ts (or .js) in the root of your project to define Middleware. For example, at the same level as pages or app, or inside src if applicable.

This will allow you to access the `/@hello` page.

![Screenshot 2023-06-04 at 10 30 02 PM](https://github.com/vercel/next.js/assets/120007119/dd36c93f-e110-4a3d-81ac-2ae52d45d229)

**Reply by litlatte** [reply]
I was think of something of these lines to but this doesnt work since it also matches /hello.
One thing you could also do is check inside the page.tsx if the route starts with @ but this means that if you have another dynamic route that doesnt start with @ you will need to have the logic inside the page

**Reply by devjiwonchoi** [reply]
@litlatte 
If that is the case, you can simply `rewrite` the path to your `/app/hello/page.tsx`.

```js
// middleware.ts

import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'

export function middleware(request: NextRequest) {
  if (request.nextUrl.pathname.startsWith('/@')) {
    return NextResponse.rewrite(new URL('/hello', request.nextUrl))
  }
}
```
This will not need a Dynamic Route segment.

![Screenshot 2023-06-04 at 11 32 13 PM](https://github.com/vercel/next.js/assets/120007119/9456a7a0-29fd-4553-85d3-dcdfe16753ca)

![Screenshot 2023-06-04 at 11 32 26 PM](https://github.com/vercel/next.js/assets/120007119/5ee8aa7d-a57f-426b-8983-390f570e25fe)

Any path starting with `/@` will display your `/hello`, and all you need to do is fetch the data you need from that route.

**Reply by litlatte** [reply]
I think my reply was not clear lemme try again
The expected behavior for what im trying to do would be this:
I visit /@elephant---> GOOD! Page load the content for elephant
I visit /elephant ---> BAD! NextJS Displays the default 404 page (without changing route)
I tried to do that with the middleware but wasnt able to achieve that
Sine both /@elephant and /elephant gave a good response.

elephant (or hello in the previous example) are both dynamic content but your points also opens an issue on the fact that you cant create even non dynamic pages that start with @


Would be nice to have something like this:

<img width="263" alt="image" src="https://github.com/vercel/next.js/assets/48565347/43d8f67a-e1d8-42a8-822e-94e945ab75ea">

**Reply by devjiwonchoi** [reply]
Since the folders starting with `@` are package scope in general and will not be recognized as `route` directories as you've noticed, the feature you are looking for is currently unavailable.

Prevent accessing directly to the `/animal` by `redirecting` to the `/404` page, since the Dynamic Route you're going to fetch on the root segment will not contain the path `/404` right? Or some other route that your root Dynamic Route will not catch.

This will cause the `Error Boundary` to activate and show the `not-found` page and will resolve your issue.

Example:

```js
// middleware.ts

import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'

export function middleware(request: NextRequest) {
  if (request.nextUrl.pathname.startsWith('/@')) {
    return NextResponse.rewrite(new URL('/animal', request.nextUrl))
  }
  if (request.nextUrl.pathname === '/animal') {
    return NextResponse.redirect(new URL('/404', request.nextUrl))
  }
}
```

https://github.com/vercel/next.js/assets/120007119/a8c77029-bb8c-4846-8cd5-22a0e109c031

Folder Structure:

![Screenshot 2023-06-05 at 12 43 41 AM](https://github.com/vercel/next.js/assets/120007119/1dba1f87-a7b4-4e3c-b964-c19717bdf5c0)

**Reply by litlatte** [reply, marked as answer]
Found a way to do it without redirect, you use [notFound](https://nextjs.org/docs/app/api-reference/functions/not-found)

For future reference if you have a prefix to your string you will have to match it manually inside the page file, if you want to throw a 404 in  case the string does not have the wanted prefix you can use the [notFound](https://nextjs.org/docs/app/api-reference/functions/not-found)

**Reply by phoneticallySAARTHaK** [reply]
@litlatte  The docs don't seem to have mention of [notFound](https://nextjs.org/docs/app/api-reference/functions/not-found) now

**Reply by litlatte** [reply]
I updated my message, thanks for pointing that out 😊

**Comment by shrekuu** [comment]
Thanks. I get this working.
Now my `/@bond` points to user profile page.

![img_v3_0274_aa9d2c78-db2e-475e-9591-94f9ccaf785g](https://github.com/vercel/next.js/assets/3136026/98ce899f-cf5d-4593-b8f0-7e1fb97b5cc5)


`/bond` points to to a specific page. Cool!!!  Thanks everybody. 
![img_v3_0274_a8532578-3523-4af5-abf7-9359e0ef14dg](https://github.com/vercel/next.js/assets/3136026/33924f59-b339-4b64-a4c8-15de959cabac)
